Add tests for audbcards.Dataset

Bahaabrougui opened the pull request:
This PR adds tests the methods inside `core.dataset.py`. It adds the following tests:
- `test_dataset()` : Tests for audbcards.Dataset properties (except `publication` and `repository_link`)
- `test_format_schemes()`: Test the `format_schemes()` method
- `test_limit_presented_samples()`: Test the `limit_presented_samples()` method

hagenw commented:
The failing test under Windows is most likely due to some undocumented change in the return value of `audeer.extract_archive()` which let's `audb` fail, see https://github.com/audeering/audeer/issues/109

We need to first fix it in `audeer` or `audb`, before you can continue here.

Bahaabrougui commented:
> The failing test under Windows is most likely due to some undocumented change in the return value of `audeer.extract_archive()` which let's `audb` fail, see [audeering/audeer#109](https://github.com/audeering/audeer/issues/109)
> 
> We need to first fix it in `audeer` or `audb`, before you can continue here.

And I was trying to debug it from here because it was confusing the fact that the pipeline wasn't failing at the early stages of this PR..

hagenw commented:
Yes, sorry for that.

I overlooked that it as we fixed it in `audb` in https://github.com/audeering/audb/pull/301, but didn't do a release afterwards. I will let you know when this is fixed and we can restart the pipeline here.

Bahaabrougui commented:
> Yes, sorry for that.
> 
> I overlooked that it as we fixed it in `audb` in [audeering/audb#301](https://github.com/audeering/audb/pull/301), but didn't do a release afterwards. I will let you know when this is fixed and we can restart the pipeline here.



> Yes, sorry for that.
> 
> I overlooked that it as we fixed it in `audb` in [audeering/audb#301](https://github.com/audeering/audb/pull/301), but didn't do a release afterwards. I will let you know when this is fixed and we can restart the pipeline here.

Alright, thanks.

Bahaabrougui commented:
The pipeline now passes after depending on audb>=1.5.1.

hagenw commented:
Maybe we still use some user cache folders?
When I execute the tests locally it fails with:
```
tmpdir = local('/tmp/pytest-of-hwierstorf/pytest-9/test_dataset0')                                                                                                                             db = name: db        
description: Example database.
source: https://github.com/audeering/audbcards
usage: unrestricted                
languages: ...: 1.0.0              
  flavor: {bit_depth: null, channels: null, format: null, mixdown: false, sampling_rate: null}                                                                                                   complete: true                                                                                                                                                                            
                                         
    def test_dataset(tmpdir, db):
        dataset = audbcards.Dataset(pytest.NAME, pytest.VERSION)       
                                                                                                                                                                                               
        # __init__                                                            
        assert dataset.name == pytest.NAME                                                                                                                                                     
        assert dataset.version == pytest.VERSION                                                                                                                                               
        assert dataset.repository == pytest.REPOSITORY
        expected_header = audb.info.header(                                    
            pytest.NAME,         
            version=pytest.VERSION,
            cache_root=CACHE,                                                  
        )
        assert str(dataset.header) == str(expected_header)
        expected_deps = audb.dependencies(
            pytest.NAME,
            version=pytest.VERSION,
            cache_root=CACHE,
        )
        pd.testing.assert_frame_equal(dataset.deps(), expected_deps())

        # archives
        print(f'{dataset.archives=}')
        print(f'{db.files=}')
>       assert dataset.archives == str(len(db.files))
E       AssertionError: assert '2' == '3'
E         - 3
E         + 2

tests/test_dataset.py:45: AssertionError
```

hagenw commented on tests/test_dataset.py:
```diff
@@ -0,0 +1,254 @@
+import matplotlib.pyplot as plt
+
+
+CACHE = './cache'
```
This might lead to the problems I experienced as the user might have different content in that folder.
One idea would be to delete that folder at the beginning of the tests, but better would be to provide the cache folder by a fixture, e.g. you could add to `conftest.py`
```python
@pytest.fixture
def cache(tmpdir, scope='function'):
    return audeer.mkdir(audeer.path(tmpdir, 'cache'))
```
and then use the fixture inside `test_dataset()` with `test_dataset(tmpdir, db, cache)`

Bahaabrougui replied:
Yes you're right I had to delete the cache everytime before I run the tests. I've applied the change and added the fixture now.

Bahaabrougui replied:
In a future PR let's handle the caching of the `Dataset` class itself in a better way but this PR as you said let's keep the current solution.

hagenw replied:
Cool, thanks.

hagenw commented:
For me locally the test fails with:

```
>       pd.testing.assert_frame_equal(dataset.deps(), expected_deps())

tests/test_dataset.py:39: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pandas/_libs/testing.pyx:52: in pandas._libs.testing.assert_almost_equal
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   AssertionError: DataFrame.index are different
E   
E   DataFrame.index values are different (40.0 %)
E   [left]:  Index(['db.files.csv', 'db.segments.csv', 'db.speaker.csv', 'data/f0.wav',
E          'data/f1.wav'],
E         dtype='object')
E   [right]: Index(['db.files.csv', 'db.segments.csv', 'db.speaker.csv', 'data/f1.wav',
E          'data/f0.wav'],
E         dtype='object')

pandas/_libs/testing.pyx:167: AssertionError
```

So I guess we need to wait until https://github.com/audeering/audb/pull/304 is finished, or we you need to adjust the tests and by first sort the dataframes based on their index.

Bahaabrougui commented:
> For me locally the test fails with:
> 
> ```
> >       pd.testing.assert_frame_equal(dataset.deps(), expected_deps())
> 
> tests/test_dataset.py:39: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> pandas/_libs/testing.pyx:52: in pandas._libs.testing.assert_almost_equal
>     ???
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> >   ???
> E   AssertionError: DataFrame.index are different
> E   
> E   DataFrame.index values are different (40.0 %)
> E   [left]:  Index(['db.files.csv', 'db.segments.csv', 'db.speaker.csv', 'data/f0.wav',
> E          'data/f1.wav'],
> E         dtype='object')
> E   [right]: Index(['db.files.csv', 'db.segments.csv', 'db.speaker.csv', 'data/f1.wav',
> E          'data/f0.wav'],
> E         dtype='object')
> 
> pandas/_libs/testing.pyx:167: AssertionError
> ```
> 
> So I guess we need to wait until [audeering/audb#304](https://github.com/audeering/audb/pull/304) is finished, or we you need to adjust the tests and by first sort the dataframes based on their index.

I don't know why it fails, I tried generating two files one short enough and the other too long to be considered as you suggested so the same file is always returned, aren't you using the old cache by chance?

hagenw commented on tests/test_dataset.py:
```diff
@@ -0,0 +1,253 @@
+
+
+def test_dataset(cache, tmpdir, db):
+    dataset = audbcards.Dataset(pytest.NAME, pytest.VERSION)
```
To make sure we do not use an existing user folder with `Dataset` during the tests as cache, I would change here to

```suggestion
    dataset_cache = audeer.mkdir(audeer.path(tmpdir, 'cache'))
    dataset = audbcards.Dataset(
        pytest.NAME,
        pytest.VERSION,
        cache_root=dataset_cache,
    )
```

hagenw commented:
> > For me locally the test fails with:
> > ```
> > >       pd.testing.assert_frame_equal(dataset.deps(), expected_deps())
> > 
> > tests/test_dataset.py:39: 
> > _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> > pandas/_libs/testing.pyx:52: in pandas._libs.testing.assert_almost_equal
> >     ???
> > _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> > 
> > >   ???
> > E   AssertionError: DataFrame.index are different
> > E   
> > E   DataFrame.index values are different (40.0 %)
> > E   [left]:  Index(['db.files.csv', 'db.segments.csv', 'db.speaker.csv', 'data/f0.wav',
> > E          'data/f1.wav'],
> > E         dtype='object')
> > E   [right]: Index(['db.files.csv', 'db.segments.csv', 'db.speaker.csv', 'data/f1.wav',
> > E          'data/f0.wav'],
> > E         dtype='object')
> > 
> > pandas/_libs/testing.pyx:167: AssertionError
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So I guess we need to wait until [audeering/audb#304](https://github.com/audeering/audb/pull/304) is finished, or we you need to adjust the tests and by first sort the dataframes based on their index.
> 
> I don't know why it fails, I tried generating two files one short enough and the other too long to be considered as you suggested so the same file is always returned, aren't you using the old cache by chance?

Yes, it was failing as I still had a folder `./cache` that was used by `Dataset`. I provided a suggestion above how to avoid this problem.

hagenw commented on tests/test_dataset.py:
```diff
@@ -0,0 +1,253 @@
+        version=pytest.VERSION,
+        cache_root=cache,
+    )
+    pd.testing.assert_frame_equal(dataset.deps(), expected_deps())
+
+    # archives
+    assert dataset.archives == str(len(db.files))
```
The number of archives per database does not need to reflect the number of media files.
We can get the correct number by checking the number of archives listed in the dependency table for media files, e.g. something like:
```suggestion
    expected_df = expected_deps()
    pd.testing.assert_frame_equal(dataset.deps(), expected_df)

    # archives
    expected_archives = str(
        len(expected_df.loc[expected_deps.media].archive.unique())
    )
    assert dataset.archives == expected_archives
```

hagenw commented on tests/test_dataset.py:
```diff
@@ -0,0 +1,253 @@
+    expected_duration = str(db.files_duration(db.files).sum())
+    assert dataset.duration == expected_duration
+
+    # example
```
Here it would be good to add a little bit more documentation, e.g.
```suggestion
    # example
    # path to audio file from database
```
and maybe state how the `expected_example` path looks like, e.g. `path/relative/to-what/media/audio.wav` (because I don't know how it should look like ;) )

hagenw commented on tests/test_dataset.py:
```diff
@@ -0,0 +1,253 @@
+                ['DE', 'EN'],
+                'speaker',
+            ],
+        ),
```
It is very hard to understand this test, maybe it is easier to provide the expected result as argument and test only a few combinations?

Bahaabrougui replied:
I've changed it accordingly and added two cases for this test

hagenw replied:
Thanks.
